pkg/index/index.go

```diff
@@ -0,0 +1,223 @@
+		if c.shardWorkspaces[shard] == nil {
+			c.shardWorkspaces[shard] = map[logicalcluster.Name]map[string]logicalcluster.Name{}
+		}
+		if c.shardWorkspaces[shard][clusterName] == nil {
+			c.shardWorkspaces[shard][clusterName] = map[string]logicalcluster.Name{}
+		}
+		c.shardWorkspaces[shard][clusterName][ws.Name] = logicalcluster.New(ws.Status.Cluster)
+	}
+}
+
+func (c *State) DeleteClusterWorkspace(shard string, ws *tenancyv1alpha1.ClusterWorkspace) {
+	if ws.Status.Phase == tenancyv1alpha1.ClusterWorkspacePhaseScheduling {
+		return
+	}
+	clusterName := logicalcluster.From(ws)
+
+	c.lock.RLock()
+	_, found := c.shardWorkspaces[shard][clusterName][ws.Name]
+	c.lock.RUnlock()
+
+	if !found {
+		return
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if _, found = c.shardWorkspaces[shard][clusterName][ws.Name]; !found {
+		return
+	}
+
+	delete(c.shardWorkspaces[shard][clusterName], ws.Name)
+	if len(c.shardWorkspaces[shard][clusterName]) == 0 {
+		delete(c.shardWorkspaces[shard], clusterName)
+	}
+	if len(c.shardWorkspaces[shard]) == 0 {
+		delete(c.shardWorkspaces, shard)
+	}
+}
+
+func (c *State) UpsertThisWorkspace(shard string, this *tenancyv1alpha1.ThisWorkspace) {
+	clusterName := logicalcluster.From(this)
+
+	c.lock.RLock()
+	got := c.logicalClusterShards[clusterName]
+	c.lock.RUnlock()
+
+	if got != shard {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		c.logicalClusterShards[clusterName] = shard
+	}
+}
+
+func (c *State) DeleteThisWorkspace(shard string, this *tenancyv1alpha1.ThisWorkspace) {
+	clusterName := logicalcluster.From(this)
+
+	c.lock.RLock()
+	got := c.logicalClusterShards[clusterName]
+	c.lock.RUnlock()
+
+	if got == shard {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		if got := c.logicalClusterShards[clusterName]; got == shard {
+			delete(c.logicalClusterShards, clusterName)
+		}
+	}
+}
+
+func (c *State) UpsertShard(shardName, baseURL string) {
+	c.lock.RLock()
+	got := c.shardBaseURLs[shardName]
+	c.lock.RUnlock()
+
+	if got != baseURL {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		c.shardBaseURLs[shardName] = baseURL
+	}
+}
+
+func (c *State) DeleteShard(shardName string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	for lc, shardName := range c.logicalClusterShards {
+		if shardName == shardName {
+			delete(c.logicalClusterShards, lc)
+		}
+	}
+	delete(c.shardWorkspaces, shardName)
+	delete(c.shardBaseURLs, shardName)
+}
+
+func (c *State) LookupShardAndCluster(path logicalcluster.Name) (string, logicalcluster.Name, bool) {
+	segments := strings.Split(path.String(), ":")
+
+	for _, rewriter := range c.rewriters {
+		segments = rewriter(segments)
+	}
+
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	// walk through index graph to find the final logical cluster and shard
+	var cluster logicalcluster.Name
+	var shard string
+	for i, s := range segments {
+		if i == 0 {
+			var found bool
+			shard, found = c.logicalClusterShards[logicalcluster.New(s)]
```

Comment:
how can this work for `root:rh:lukasz:this` workspace?

Reply:
`s` in that case points to `root`

Reply:
yes, the first segment must be a logical cluster.